tests/rust/safe_abstraction/mutex_u32.rs

```diff
@@ -0,0 +1,444 @@
+#![feature(negative_impls)]
+#![allow(dead_code)]
+
+/*
+About the definitions:
+    Since in the proof of `MutexU32::lock` we need to close the `frac_borrow`,
+    in the `sys::locks::Mutex` interface; i.e. `SysMutex` fractions required by `sys::locks::Mutex::lock` to lock the mutex should be turened back from the method,
+    we cannot keep track of `SysMutex` fractions used to get the lock. So we adapth an interface for `sys::locks::Mutex` such that it does not care
+    if the mutex is locked or not.
+*/
+
+/*@
+// dummy
+pred sys::locks::Mutex_own(t: thread_id_t, m: *u32);
+pred sys::locks::Mutex_share(k: lifetime_t, t: thread_id_t, l: *sys::locks::Mutex;) = true;
+lem sys::locks::Mutex_share_mono(k: lifetime_t, k1: lifetime_t, t: thread_id_t, l: *sys::locks::Mutex)
+req lifetime_inclusion(k1, k) == true &*& [_]sys::locks::Mutex_share(k, t, l);
+ens [_]sys::locks::Mutex_share(k1, t, l);
+{}
+lem sys::locks::Mutex_share_full(k: lifetime_t, t: thread_id_t, l: *sys::locks::Mutex)
+req full_borrow(k, sys::locks::Mutex_full_borrow_content(t, l)) &*& [?q]lifetime_token(k);
+ens [_]sys::locks::Mutex_share(k, t, l) &*& [q]lifetime_token(k);
+{
+    leak full_borrow(_, _);
+}
+// ymmud
+
+pred SysMutex(m: sys::locks::Mutex; P: pred());
+pred SysMutex_share(l: *sys::locks::Mutex; P: pred());
+lem SysMutex_share_full(l: *sys::locks::Mutex);
+    req *l |-> ?m &*& SysMutex(m, ?P);
+    ens SysMutex_share(l, P);
+lem SysMutex_end_share(l: *sys::locks::Mutex);
+   req SysMutex_share(l, ?P);
+   ens *l |-> ?m &*& SysMutex(m, P);
+
+pred SysMutex::new_ghost_arg(P: pred()) = true;
```

Comment:
The modern way is to just use `exists`. (I created `threading.h` before I had thought of `exists` as a general pattern.)

Reply:
To have the spec for `sys::locks::Mutex::new` like `req exists(?P) &*& P();` causes the following error:
Variable P is not of predicate type.
It seems the type inference sees `P` as a fixpoint that returns a unit.

Reply:
To work around this limitation, write `req exists::<pred()>(?P) &*& P();`.